src/ai/ai.cpp

```diff
@@ -890,66 +908,74 @@ void update_ai_econ_construction(sys::state& state) {
 		auto rules = n.get_combined_issue_rules();
 
 		if((rules & issue_rule::expand_factory) != 0 || (rules & issue_rule::build_factory) != 0) {
-			static::std::vector<dcon::factory_type_id> desired_types;
-			desired_types.clear();
-			get_desired_factory_types(state, n, desired_types);
-
-			// desired types filled: try to construct or upgrade
-			if(!desired_types.empty()) {
-				static std::vector<dcon::state_instance_id> ordered_states;
-				ordered_states.clear();
-				for(auto si : n.get_state_ownership()) {
-					if(si.get_state().get_capital().get_is_colonial() == false)
-						ordered_states.push_back(si.get_state().id);
-				}
-				std::sort(ordered_states.begin(), ordered_states.end(), [&](auto a, auto b) {
-					auto apop = state.world.state_instance_get_demographics(a, demographics::total);
-					auto bpop = state.world.state_instance_get_demographics(b, demographics::total);
-					if(apop != bpop)
-						return apop > bpop;
-					else
-						return a.index() < b.index();
-				});
-				if((rules & issue_rule::build_factory) == 0 && (rules & issue_rule::expand_factory) != 0) { // can't build -- by elimination, can upgrade
-					for(auto si : ordered_states) {
-						if(max_projects <= 0)
-							break;
-
-						auto pw_num = state.world.state_instance_get_demographics(si,
-								demographics::to_key(state, state.culture_definitions.primary_factory_worker));
-						auto pw_employed = state.world.state_instance_get_demographics(si,
-								demographics::to_employment_key(state, state.culture_definitions.primary_factory_worker));
+			// prepare a list of states
+			static std::vector<dcon::state_instance_id> ordered_states;
+			ordered_states.clear();
+			for(auto si : n.get_state_ownership()) {
+				if(si.get_state().get_capital().get_is_colonial() == false)
+					ordered_states.push_back(si.get_state().id);
+			}
+			std::sort(ordered_states.begin(), ordered_states.end(), [&](auto a, auto b) {
+				auto apop = state.world.state_instance_get_demographics(a, demographics::total);
+				auto bpop = state.world.state_instance_get_demographics(b, demographics::total);
+				if(apop != bpop)
+					return apop > bpop;
+				else
+					return a.index() < b.index();
+			});
 
-						if(pw_employed >= pw_num && pw_num > 0.0f)
-							continue; // no spare workers
+			// try to upgrade factories first:
+			if((rules & issue_rule::build_factory) == 0 && (rules & issue_rule::expand_factory) != 0) { // can't build -- by elimination, can upgrade
```

Comment:
This condition will only pass if the nation does not have the ability to build new factories. IIRC your rational for moving this loop up was to prioritize upgrading, even for nations that could build, so you may want to remove the first part of the condition.